### PR TITLE
i#1569 AArch64: Make encodings easier to specify.

### PR DESCRIPTION
**Pull request by egrimley**
In the operand definitions, in codec.txt, '?' indicates a bit that
the operand decoder/encoder may consult as it decodes or encodes the
bits labelled 'x'. An operand encoder only receives the bits
labelled '?'; an operand decoder only receives the bits labelled '?'
or 'x'. An operand encoder may generate a bit that is read by another
operand encoder, provided there is no cyclic dependency. This change
makes it easier to specify some encodings, and to detect errors.

A few changes are made exploiting the new facilities:

- memvrpost is removed: memvr is sufficient
- some patterns for SIMD load/store structure are combined
- some patterns for data processing with shifted register are combined
  (for which "imm6" is modified: it now checks bit 31 for consistency)

Change-Id: Ic5ef256af143239521c4cea86631dabebd763ea1

### CODE REVIEW

**Comment by egrimley** [comment]
AppVeyor failure looks like #2246.

File: core/arch/aarch64/codec.c [hunk 1]
```diff
@@ -1128,12 +1128,16 @@ encode_opnd_imm5(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
 static inline bool
 decode_opnd_imm6(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
 {
+    if (!TEST(1U << 31, enc) && TEST(1U << 15, enc))
+        return false;
```
Comment (derekbruening) [inline core/arch/aarch64/codec.c]: Not sure I see where this is coming from.  This is a bug fix?  Should be mentioned in the commit message?  Same below.

File: core/arch/aarch64/codec.py [hunk 2]
```diff
@@ -55,33 +71,35 @@ def gen1(c, opndsgen):
...
+            c += ['    instr_set_opcode(instr, opcode);',
+                  '    instr_set_num_opnds(dcontext, instr, %d, %d);' %
+                  (len(dsts), len(srcs))]
+            c += ['    instr_set_dst(instr, %d, dst%d);' % (i, i)
+                  for i in range(len(dsts))]
+            c += ['    instr_set_src(instr, %d, src%d);' % (i, i)
+                  for i in range(len(srcs))]
+            c += ['    return true;',
+                  '}',
+                  '']
```
Comment (derekbruening) [inline core/arch/aarch64/codec.py]: This is all just cosmetic?

File: core/arch/aarch64/codec.py [hunk 3]
```diff
@@ -98,39 +116,63 @@ def gen(c, pats, depth):
...
+                pats0 += [p]
             if (1 << best_b) & (f | v):
-                pats1.append(p)
+                pats1 += [p]
         gen(c, pats0, depth + 1)
-        c.append('%s} else {' % indent)
+        c += ['%s} else {' % indent]
         gen(c, pats1, depth + 1)
-        c.append('%s}' % indent)
+        c += ['%s}' % indent]
```
Comment (derekbruening) [inline core/arch/aarch64/codec.py]: All of these s/append/+=/ seem to confuse the issue here: unrelated to the content changes, right?  Maybe better to be a separate CL.

File: core/arch/aarch64/codec.c (hunk 1 above)
Comment (egrimley) [inline core/arch/aarch64/codec.c]: It has to do with "data processing with shifted register". I've added a parenthesis to the draft log message ("comment") above.

File: core/arch/aarch64/codec.py (hunk 3 above)
Comment (egrimley) [inline core/arch/aarch64/codec.py]: I'll undo the cosmetic changes for now.

**Pull request merged**